tile.1
.\" 
.TH "tile" "1" "0.7.2" "Greg Schenzel" ""
.SH "NAME"
tile \- arrange X11 windows on the desktop
.SH "SYNOPSIS"
.B tile
[\fB\-\-verbose\fR]
[\fB\-h\fR
| \fB\-v\fR]
[\fB\-p\fIprofile\fR]
[\fB\-m\fR
| \fB\-w\fR]
[\fB\-\-help\fR]
.SH "DESCRIPTION"
This is a utility to tile windows on the desktop. It supports user\-
configurable window manager profiles to avoid screen objects and can
tile in a horizontal or vertical fashion. It works well with MWM,
Oroborus and some multi\-desktop WMs like FluxBox (given the \-m
option) or GNOME/Net-WM compliant WMs (given the \-w option). It is
somewhat tested with GNOME, and not tested with
KDE; good luck. If testing out new territory, I encourage you to
submit the wmprofiles and "avoid" rc data that you use.
.SH "OPTIONS"
.TP 
.B \-v,\-\-vertical
vertical placement
.TP 
.B \-h,\-\-horizontal
horizontal placement
.TP
.B \-w,\-\-net-wm
enable multi\-desktop operation (GNOME, NET-WM compatible, etc)
.TP 
.B \-m,\-\-win\-workspace
enable multi\-desktop operation (oroborus, fluxbox, etc)
.TP 
\fB\-p\fIprofn\fR,\fB\-\-profile=\fIprofn\fR
window manager profile name, from $PREFIX/share/tile/wmprofiles or
$HOME/.tile/wmprofiles.
.TP 
.B \-\-verbose
verbose
.TP 
.B \-\-help
help

.SH "AUTHOR"
This software written by Greg Schenzel. Please mail bug reports, feature requests, and patches to inittab@unixdev.net.


ChangeLog
0.7.4

  * If run as "iconify", tile will minimize all windows (show desktop for MS users)


0.7.3

  * packaging fixes (for NetBSD)


0.7.2

  * streamlining of program output
  * --dump option added
  * lintian: INSTALL file not packaged


0.7.1

  * CLI profile selection fix
  * Corrected manpage data


0.7.0-1

  * wmavoid file deprecated in favor of rc file
  * Split up source code
  * now uses GNU getopt-long


0.6.4-3

  * wmavoid syntax changed to allow specification of Atom (XA_STRING)


0.6.4-2

  * Added .desktop files for GNOME 2


0.6.4-1

  * New upstream version.
  * wmavoid configuration files (local or system) lists window names to ignore (use xprop | grep WM_NAME)
  * NET-WM support added (-w, very similar to -m)
  * Corrected manpage and other documentation


0.6.2-3

  * more lintian fixes
  * manpage written
  
   
0.6.2-2

  * lintian source fixes


0.6.2

  * New upstream version.
  * New maintainer.



tile.c
/*
 * Copyright (c) 2005 Greg Schenzel
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <X11/X.h>
#include <X11/Xlib.h>
#include <X11/Xatom.h>
#include <X11/Xutil.h>
#include <X11/Intrinsic.h>
#include <stdio.h>
#include <stdlib.h>
#include "tilerc.h"
#include "tile.h"
#include "x11.h"
#include "cmdline.h"
#include <libgen.h>

/** Entry point. */
int main(int argc, char **argv)
{
    Display *dsp;
    Window win;
    char * invocation = basename(argv[0]);
    struct tile_config *tileconf;
    struct tile_state *tilestate;

    tileconf = (struct tile_config *) malloc(sizeof(struct tile_config));
    tilestate = (struct tile_state *) malloc(sizeof(struct tile_state));

    cmdline_parser(argc, argv, tileconf);

    //signal(11, catch_sig); //catch SIGSEGV

    if ((dsp = XOpenDisplay("")) == NULL) {
        fputs("ERROR: Could not connect to X11 Server.\n", stderr);
        exit(1);
    }

    win = DefaultRootWindow(dsp);

    tilestate->minimize = strcmp(invocation, "iconify") == 0;
    tilestate->numwins2 = 0;
    tilestate->numwins = 0;
    tilestate->padleft = -1;
    tilestate->curdesk = 0;

    if (tileconf->wmulti || tileconf->multi) {
        Atom real_type;
        Atom work;
        int format;
        unsigned long n, extra;
        int status;
        long *p = 0;
        work =
            XInternAtom(dsp,
                        tileconf->
                        multi ? "_WIN_WORKSPACE" : "_NET_CURRENT_DESKTOP",
                        False);
        status =
            XGetWindowProperty(dsp, win, work, 0L, 1L, False, XA_CARDINAL,
                               &real_type, &format, &n, &extra,
                               (unsigned char **) &p);
        if (!status) {
            if (p) {
                tilestate->curdesk = p[0];
            }
        }
    }

    parseProfiles(tileconf);

    tilestate->scrwidth =
        getGeom(dsp, win, GWIDTH) - (tileconf->scpadl + tileconf->scpadr);
    tilestate->scrheight =
        getGeom(dsp, win, GHEIGHT) - (tileconf->scpadt + tileconf->scpadb);

    if (tileconf->verbose >= GT_ACTION)
        puts("recurseWindow (pass 1): find windows for count");
    recurseWindow(dsp, win, tileconf, tilestate);

    if (tileconf->verbose >= GT_ACTION)
        puts("moveResizeWindow (pass 2): move and resize windows");
    moveResizeWindow(dsp, win, tileconf, tilestate);

    return 0;
}

/** Catch a kill signal. */
void catch_sig(int sig_num)
{
    /* re-set the signal handler again to catch_int, for next time */
    //    signal(SIGINT, catch_int);
    puts("ERROR: Exiting, caught signal!");
    exit(1);
}

/** Dump the program configuration and exit. */
void dumpConfig(struct tile_config *tileconf)
{
    printf("Orientation: %s\n",
           tileconf->orientation == VERTICAL ? "vertical" : "horizontal");
    printf("Verbosity Level: %d\n", tileconf->verbose);
    printf("Multi-Desk Workspace: %s\n",
           tileconf->multi ? "True" : "False");
    printf("Multi-Desk Net-WM: %s\n", tileconf->wmulti ? "True" : "False");
    printf("WM Profile: %s\n", tileconf->profile_name);
    exit(0);
}
